File: lib/karakeep.ts
export type KarakeepBookmark = {
  id: string;
  title: string;
  excerpt: string;
  link: string;
  created: string;
  tags: string[];
  cover: string | null;
};

type RawTag = { id: string; name: string; attachedBy?: string };

type RawBookmark = {
  id: string;
  createdAt: string;
  title: string | null;
  tags: RawTag[];
  content: {
    type: string;
    url?: string;
    title?: string | null;
    description?: string | null;
    imageUrl?: string | null;
  };
};

type BookmarksResponse = {
  bookmarks: RawBookmark[];
  nextCursor: string | null;
};

const DEFAULT_HOST = "https://bookmarks.furkanunsalan.dev";

export const BOOKMARK_LISTS = [
  {
    id: "oalr80pga230xqpupf836v31",
    title: "Posts",
    description:
      "An archive for the posts that I've bookmarked and read along the way.",
  },
  {
    id: "zy0ciwhnnm9lr113ze8ex51a",
    title: "Videos",
    description:
      "The videos I've watched and found useful without any categorizations",
  },
] as const;

function host(): string {
  return process.env.KARAKEEP_API_URL || DEFAULT_HOST;
}

function authHeaders(): HeadersInit {
  const key = process.env.KARAKEEP_API_KEY;
  if (!key) throw new Error("Karakeep API key is not configured");
  return {
    Authorization: `Bearer ${key}`,
    "Content-Type": "application/json",
  };
}

function normalize(raw: RawBookmark): KarakeepBookmark {
  const c = raw.content || ({} as RawBookmark["content"]);
  return {
    id: raw.id,
    title: raw.title || c.title || c.url || "Untitled",
    excerpt: c.description || "",
    link: c.url || "#",
    created: raw.createdAt,
    tags: (raw.tags || []).map((t) => t.name),
    cover: c.imageUrl || null,
  };
}

async function fetchBookmarksPage(
  path: string,
  cursor: string | null,
  limit: number,
): Promise<BookmarksResponse> {
  const url = new URL(`${host()}${path}`);
  url.searchParams.set("limit", String(limit));
  url.searchParams.set("includeContent", "false");
  if (cursor) url.searchParams.set("cursor", cursor);

  const res = await fetch(url.toString(), {
    headers: authHeaders(),
    cache: "no-store",
  });
  if (!res.ok) {
    throw new Error(`Karakeep request failed: ${res.status}`);
  }
  return (await res.json()) as BookmarksResponse;
}

export async function getKarakeepBookmarks(
  listId: string,
  cursor: string | null = null,
  limit: number = 50,
) {
  const data = await fetchBookmarksPage(
    `/api/v1/lists/${listId}/bookmarks`,
    cursor,
    limit,
  );
  return {
    items: data.bookmarks.map(normalize),
    nextCursor: data.nextCursor,
  };
}

export async function getKarakeepListCount(listId: string): Promise<number> {
  let cursor: string | null = null;
  let total = 0;
  do {
    const data = await fetchBookmarksPage(
      `/api/v1/lists/${listId}/bookmarks`,
      cursor,
      100,
    );
    total += data.bookmarks.length;
    cursor = data.nextCursor;
  } while (cursor);
  return total;
}

export async function getKarakeepLatest(): Promise<KarakeepBookmark | null> {
  const data = await fetchBookmarksPage(`/api/v1/bookmarks`, null, 1);
  const first = data.bookmarks[0];
  return first ? normalize(first) : null;
}

export async function getKarakeepLatestFromLists(
  listIds: string[],
): Promise<KarakeepBookmark | null> {
  const heads = await Promise.all(
    listIds.map(async (id) => {
      const data = await fetchBookmarksPage(
        `/api/v1/lists/${id}/bookmarks`,
        null,
        1,
      );
      return data.bookmarks[0] ? normalize(data.bookmarks[0]) : null;
    }),
  );
  const valid = heads.filter((b): b is KarakeepBookmark => b !== null);
  if (valid.length === 0) return null;
  return valid.sort(
    (a, b) => new Date(b.created).getTime() - new Date(a.created).getTime(),
  )[0];
}

export async function getKarakeepLast24hCount(): Promise<number> {
  const cutoff = Date.now() - 24 * 60 * 60 * 1000;
  let cursor: string | null = null;
  let count = 0;
  while (true) {
    const data: BookmarksResponse = await fetchBookmarksPage(
      `/api/v1/bookmarks`,
      cursor,
      50,
    );
    let stop = false;
    for (const b of data.bookmarks) {
      if (new Date(b.createdAt).getTime() >= cutoff) {
        count += 1;
      } else {
        stop = true;
        break;
      }
    }
    if (stop || !data.nextCursor) break;
    cursor = data.nextCursor;
  }
  return count;
}

File: terminal/internal/data/karakeep.go
package data

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"sort"
	"time"
)

type Bookmark struct {
	ID      string
	Title   string
	Excerpt string
	Link    string
	Created string
	Tags    []string
	List    string
}

// Mirrors BOOKMARK_LISTS in lib/karakeep.ts: the site only publishes these two
// curated lists, so the terminal must not expose the whole account either.
var bookmarkLists = []struct {
	ID    string
	Title string
}{
	{"oalr80pga230xqpupf836v31", "Posts"},
	{"zy0ciwhnnm9lr113ze8ex51a", "Videos"},
}

const defaultKarakeepHost = "https://bookmarks.furkanunsalan.dev"

func karakeepHost() string {
	if h := os.Getenv("KARAKEEP_API_URL"); h != "" {
		return h
	}
	return defaultKarakeepHost
}

func karakeepKey() string { return os.Getenv("KARAKEEP_API_KEY") }

type karakeepPage struct {
	Bookmarks []struct {
		ID        string `json:"id"`
		CreatedAt string `json:"createdAt"`
		Title     string `json:"title"`
		Tags      []struct {
			Name string `json:"name"`
		} `json:"tags"`
		Content struct {
			URL         string `json:"url"`
			Title       string `json:"title"`
			Description string `json:"description"`
		} `json:"content"`
	} `json:"bookmarks"`
	NextCursor string `json:"nextCursor"`
}

func fetchKarakeepList(ctx context.Context, listID string, limit int) ([]Bookmark, error) {
	u, err := url.Parse(fmt.Sprintf("%s/api/v1/lists/%s/bookmarks", karakeepHost(), listID))
	if err != nil {
		return nil, err
	}
	q := u.Query()
	q.Set("limit", fmt.Sprintf("%d", limit))
	q.Set("includeContent", "false")
	u.RawQuery = q.Encode()

	req, _ := http.NewRequestWithContext(ctx, "GET", u.String(), nil)
	req.Header.Set("Authorization", "Bearer "+karakeepKey())
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 10 * time.Second}
	res, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("karakeep status %d", res.StatusCode)
	}

	var page karakeepPage
	if err := json.NewDecoder(res.Body).Decode(&page); err != nil {
		return nil, err
	}

	out := make([]Bookmark, 0, len(page.Bookmarks))
	for _, raw := range page.Bookmarks {
		title := raw.Title
		if title == "" {
			title = raw.Content.Title
		}
		if title == "" {
			title = raw.Content.URL
		}
		if title == "" {
			title = "Untitled"
		}
		link := raw.Content.URL
		if link == "" {
			link = "#"
		}
		tags := make([]string, 0, len(raw.Tags))
		for _, t := range raw.Tags {
			tags = append(tags, t.Name)
		}
		out = append(out, Bookmark{
			ID:      raw.ID,
			Title:   title,
			Excerpt: raw.Content.Description,
			Link:    link,
			Created: raw.CreatedAt,
			Tags:    tags,
		})
	}
	return out, nil
}

func karakeepCreatedAt(b Bookmark) time.Time {
	ts, err := time.Parse(time.RFC3339, b.Created)
	if err != nil {
		return time.Time{}
	}
	return ts
}

// LoadBookmarks fetches the latest N bookmarks across the published lists.
func LoadBookmarks(ctx context.Context, limit int) ([]Bookmark, error) {
	if karakeepKey() == "" {
		return nil, fmt.Errorf("KARAKEEP_API_KEY not set")
	}
	if limit <= 0 || limit > 50 {
		limit = 50
	}

	var all []Bookmark
	seen := map[string]bool{}
	var firstErr error
	for _, list := range bookmarkLists {
		items, err := fetchKarakeepList(ctx, list.ID, limit)
		if err != nil {
			if firstErr == nil {
				firstErr = err
			}
			continue
		}
		for _, b := range items {
			if seen[b.ID] {
				continue
			}
			seen[b.ID] = true
			b.List = list.Title
			all = append(all, b)
		}
	}
	if len(all) == 0 && firstErr != nil {
		return nil, firstErr
	}

	sort.SliceStable(all, func(i, j int) bool {
		return karakeepCreatedAt(all[i]).After(karakeepCreatedAt(all[j]))
	})
	if len(all) > limit {
		all = all[:limit]
	}
	return all, nil
}
